### PR TITLE
WIP: Addressing multiple issue in one

### PR DESCRIPTION
**Pull request by dteague**
Probably not the best practice, but I'm addressing a bunch of issues in one PR. I'm trying to refactor the code in general, but I'll try to keep issues isolated into different commits:

## Issues
~Issue #31 is covered with commit 701df45~
~Issue #33 is spread out over 160c925 and fb256af (mainly `ScaleWeightGroup` and `LHEWeightHelper`, hopefully will expand)~
Issue #29 TBD (hopefully after testing commit 160c925)
~Issue #28 TBD Hoping to fix this after basic refactoring~ Needs to be checked that functionality is still there/expand it, but basics now exist

### CODE REVIEW

**Comment by dteague** [comment]
New commits fix some errors (some introduced in my previous commits, one big one in the pdf weight splitting) and start adding more debug/error throwing. Working towards testing all the files in the google sheet and making sure it runs or fails in a controlled way right now

File: GeneratorInterface/Core/src/WeightHelper.cc [hunk 1]
```diff
@@ -85,45 +78,45 @@ namespace gen {
...
-    } catch (std::invalid_argument& e) {
-      scaleGroup.setIsWellFormed(false);
     }
+
     if (scaleGroup.lhaid() == -1) {
       std::string lhaidText = searchAttributes("pdf", weight);
       try {
         scaleGroup.setLhaid(std::stoi(lhaidText));
-      } catch (std::invalid_argument& e) {
+      } catch (...) {
```
Comment (kdlong) [inline GeneratorInterface/Core/src/WeightHelper.cc:115]: I would think it's better to just catch the exception you expect. Why is it needed to be so inclusive?

File: GeneratorInterface/Core/src/WeightHelper.cc [hunk 2]
```diff
@@ -85,45 +78,45 @@ namespace gen {
...
-        std::string dynType = searchAttributes("dyn_name", weight);
+    if (dynNumText.empty()) {
+      scaleGroup.setMuRMuFIndex(weight.index, weight.id, muR, muF);
+    } else {
+      std::string dynType = searchAttributes("dyn_name", weight);
+      try {
         int dynNum = std::stoi(dynNumText);
-        scaleGroup.setMuRMuFIndex(weight.index, weight.id, muR, muF, dynNum, dynType);
+        scaleGroup.setDyn(weight.index, weight.id, muR, muF, dynNum, dynType);
+      } catch (...) {
```
Comment (kdlong) [inline GeneratorInterface/Core/src/WeightHelper.cc:104]: Can we catch the specific exception you expect to get?

File: GeneratorInterface/Core/src/WeightHelper.cc [hunk 3]
```diff
@@ -85,45 +78,45 @@ namespace gen {
...
-      scaleGroup.setIsWellFormed(false);
     }
+
     if (scaleGroup.lhaid() == -1) {
       std::string lhaidText = searchAttributes("pdf", weight);
       try {
         scaleGroup.setLhaid(std::stoi(lhaidText));
-      } catch (std::invalid_argument& e) {
+      } catch (...) {
         scaleGroup.setLhaid(-2);
```
Comment (kdlong) [inline GeneratorInterface/Core/src/WeightHelper.cc:116]: What is special about -2?

File: SimDataFormats/GeneratorProducts/interface/ScaleWeightGroupInfo.h [hunk 4]
```diff
@@ -12,65 +12,79 @@ namespace gen {
     std::vector<size_t> muIndices_;
     bool containsCentral_ = false;
     int lhaid_ = -1;
+    bool hasAllWeights = false;
```
Comment (kdlong) [inline SimDataFormats/GeneratorProducts/interface/ScaleWeightGroupInfo.h:15]: Why is this needed in addition to isWellFormed?

File: SimDataFormats/GeneratorProducts/interface/ScaleWeightGroupInfo.h [hunk 5]
```diff
@@ -12,65 +12,79 @@ namespace gen {
...
+      isWellFormed_ = true;
     }
     ScaleWeightGroupInfo(std::string header) : ScaleWeightGroupInfo(header, header) {}
     ScaleWeightGroupInfo(const ScaleWeightGroupInfo& other) { copy(other); }
     virtual ~ScaleWeightGroupInfo() override {}
     void copy(const ScaleWeightGroupInfo& other);
     virtual ScaleWeightGroupInfo* clone() const override;
     bool containsCentralWeight() const { return containsCentral_; }
+    void addContainedId(int globalIndex, std::string id, std::string label, float muR, float muF);
+    bool isWellFormed() { return isWellFormed_ && hasAllWeights; }
```
Comment (kdlong) [inline SimDataFormats/GeneratorProducts/interface/ScaleWeightGroupInfo.h:51]: I think this function should just return isWellFormed_, and it needs to be set appropriately elsewhere. Otherwise it's really misleading

**Comment by kdlong** [comment]
@dteague did you see the comments I made?

File: GeneratorInterface/Core/src/WeightHelper.cc (hunk 3 above)
Comment (dteague) [reply GeneratorInterface/Core/src/WeightHelper.cc:116]: To be honest, I don't remember exactly now, the blame doesn't give a lot of help. I think the intent was to distinguish the default value (-1) from an error in the LHAid text (ie pdf info in the xml but its not an integer). Considering I can't fully remember/find the intent of this code means it can be cleaned up... If the distinction between the two doesn't matter, we can just set it to -1 like the default value, else I can make some sort of enum to clarify this distinction.

File: SimDataFormats/GeneratorProducts/interface/ScaleWeightGroupInfo.h (hunk 4 above)
Comment (dteague) [reply SimDataFormats/GeneratorProducts/interface/ScaleWeightGroupInfo.h:15]: `isWellFormed_` defaults to true and is only set to false if one of the weights has a bad value for muR or muF or any other formatting issue. I want to keep that functionality, but also have a value that updates whether or not the scale variation has all 9 weights. That's why the extra weight. I'm not sure if that information is wanted/needed. A work around would be to have the checking if all the 9 weights exist in the `isWellFormed()`, but that would require iterating through an array every time

File: SimDataFormats/GeneratorProducts/interface/ScaleWeightGroupInfo.h (hunk 5 above)
Comment (dteague) [reply SimDataFormats/GeneratorProducts/interface/ScaleWeightGroupInfo.h:51]: See above comment. I guess the large question is whether a scale group having all 9 variations should be under the purview of the `isWellFormed` function. Because the `muIndices_` vector is sized to have 9 indices, its not trivial to check if all 9 indices have not -1 values

File: GeneratorInterface/Core/src/WeightHelper.cc (hunk 1 above)
Comment (dteague) [reply GeneratorInterface/Core/src/WeightHelper.cc:115]: I think for this comment and the next, the errrors only come form the `std::stoi`, so I can be more specific. I think the lack of specificity was more my laziness...

File: GeneratorInterface/Core/src/WeightHelper.cc (hunk 2 above)
Comment (dteague) [reply GeneratorInterface/Core/src/WeightHelper.cc:104]: See above

File: GeneratorInterface/Core/src/WeightHelper.cc (hunk 1 above)
Comment (kdlong) [reply GeneratorInterface/Core/src/WeightHelper.cc:115]: I imagine this will come up in the cmssw review, better to address it now

File: GeneratorInterface/Core/src/WeightHelper.cc (hunk 3 above)
Comment (kdlong) [reply GeneratorInterface/Core/src/WeightHelper.cc:116]: yeah it's confusing like this, lets revert and fix again if need be

File: SimDataFormats/GeneratorProducts/interface/ScaleWeightGroupInfo.h (hunk 4 above)
Comment (kdlong) [reply SimDataFormats/GeneratorProducts/interface/ScaleWeightGroupInfo.h:15]: My feeling is that if there is an isWellFormed member variable and an isWellFormed member function, the user will assume that the function returns the variable.

Comment (dteague) [reply SimDataFormats/GeneratorProducts/interface/ScaleWeightGroupInfo.h:15]: fair enough. Though I'll have to think through the logistics. The trouble comes from we want the `isWellFormed_` to be false always when there is a malformed variable, to be false when there are not 9 weights, but to flip to true if there are 9 weights, non of which are malformed. I can change to have a private variable `isMalformed_` which is only used in `setMuRMuFIndex` for deciding whether to set `isWellFormed_` to true if there are 9 weights and non are malformed

File: SimDataFormats/GeneratorProducts/interface/ScaleWeightGroupInfo.h (hunk 5 above)
Comment (kdlong) [reply SimDataFormats/GeneratorProducts/interface/ScaleWeightGroupInfo.h:51]: I think isWellFormed should indicate that there are 9 properly parsed variations.
